File: internal/awiki/wiki/frontmatter.go
package wiki

import "strings"

type FrontMatter struct {
	Present    bool
	BodyOffset int
	Title      string
	Aliases    []string
}

type textLine struct {
	text  string
	start int
	end   int
}

func ParseFrontMatter(content string) FrontMatter {
	lines := scanLines(content)
	endIndex, ok := frontMatterEndIndex(lines)
	if !ok {
		return FrontMatter{}
	}

	fm := FrontMatter{
		Present:    true,
		BodyOffset: lines[endIndex].end,
	}

	for i := 1; i < endIndex; i++ {
		line := trimLine(lines[i].text)
		if line == "" || startsIndented(line) {
			continue
		}

		key, value, ok := splitKeyValue(line)
		if !ok {
			continue
		}

		switch key {
		case "title":
			fm.Title = trimYAMLScalar(value)
		case "aliases":
			aliases, next := parseAliases(lines, i, endIndex, value)
			fm.Aliases = appendDedup(fm.Aliases, aliases...)
			i = next
		}
	}

	return fm
}

func UpdateFrontMatterTitle(content, oldTitle, newTitle string) (string, bool) {
	lines := scanLines(content)
	endIndex, ok := frontMatterEndIndex(lines)
	if !ok {
		return content, false
	}

	index, value, found := findTopLevelKey(lines, endIndex, "title")
	if !found || trimYAMLScalar(value) != oldTitle {
		return content, false
	}

	lines[index].text = replaceLineText(lines[index].text, "title: "+formatScalarLike(value, newTitle))
	return joinLines(lines), true
}

func parseAliases(lines []textLine, start, end int, value string) (aliases []string, next int) {
	trimmed := strings.TrimSpace(value)
	next = start
	if trimmed == "" {
		for i := start + 1; i < end; i++ {
			line := trimLine(lines[i].text)
			if line == "" {
				next = i
				continue
			}
			if !startsIndented(trimLinePreserveIndent(lines[i].text)) {
				break
			}

			item := strings.TrimSpace(line)
			if !strings.HasPrefix(item, "- ") {
				next = i
				continue
			}
			aliases = append(aliases, trimYAMLScalar(strings.TrimSpace(strings.TrimPrefix(item, "- "))))
			next = i
		}
		return aliases, next
	}

	if strings.HasPrefix(trimmed, "[") && strings.HasSuffix(trimmed, "]") {
		return parseInlineList(trimmed), next
	}

	return []string{trimYAMLScalar(trimmed)}, next
}

func parseInlineList(value string) []string {
	inner := strings.TrimSpace(value[1 : len(value)-1])
	if inner == "" {
		return nil
	}

	var (
		items   []string
		current strings.Builder
		quote   rune
	)

	for _, r := range inner {
		switch {
		case quote == 0 && (r == '\'' || r == '"'):
			quote = r
			current.WriteRune(r)
		case quote != 0 && r == quote:
			quote = 0
			current.WriteRune(r)
		case quote == 0 && r == ',':
			items = append(items, trimYAMLScalar(strings.TrimSpace(current.String())))
			current.Reset()
		default:
			current.WriteRune(r)
		}
	}

	if current.Len() > 0 {
		items = append(items, trimYAMLScalar(strings.TrimSpace(current.String())))
	}

	return items
}

func splitKeyValue(line string) (key, value string, ok bool) {
	idx := strings.Index(line, ":")
	if idx == -1 {
		return "", "", false
	}
	return strings.TrimSpace(line[:idx]), strings.TrimSpace(line[idx+1:]), true
}

func trimYAMLScalar(value string) string {
	value = strings.TrimSpace(value)
	if len(value) >= 2 {
		if value[0] == '"' && value[len(value)-1] == '"' {
			return value[1 : len(value)-1]
		}
		if value[0] == '\'' && value[len(value)-1] == '\'' {
			return value[1 : len(value)-1]
		}
	}
	return value
}

func formatScalarLike(originalValue, replacement string) string {
	trimmed := strings.TrimSpace(originalValue)
	if len(trimmed) >= 2 {
		if trimmed[0] == '"' && trimmed[len(trimmed)-1] == '"' {
			return `"` + replacement + `"`
		}
		if trimmed[0] == '\'' && trimmed[len(trimmed)-1] == '\'' {
			return `'` + replacement + `'`
		}
	}
	return replacement
}

func scanLines(content string) []textLine {
	if content == "" {
		return nil
	}

	var lines []textLine
	start := 0
	for start < len(content) {
		end := start
		for end < len(content) && content[end] != '\n' {
			end++
		}
		if end < len(content) {
			end++
		}
		lines = append(lines, textLine{
			text:  content[start:end],
			start: start,
			end:   end,
		})
		start = end
	}

	return lines
}

func joinLines(lines []textLine) string {
	var b strings.Builder
	for _, line := range lines {
		b.WriteString(line.text)
	}
	return b.String()
}

func trimLine(line string) string {
	return strings.TrimRight(line, "\r\n")
}

func trimLinePreserveIndent(line string) string {
	return strings.TrimRight(line, "\r\n")
}

func frontMatterEndIndex(lines []textLine) (int, bool) {
	if len(lines) == 0 || trimLine(lines[0].text) != "---" {
		return -1, false
	}
	for i := 1; i < len(lines); i++ {
		if trimLine(lines[i].text) == "---" {
			return i, true
		}
	}
	return -1, false
}

func findTopLevelKey(lines []textLine, endIndex int, wanted string) (index int, value string, ok bool) {
	for i := 1; i < endIndex; i++ {
		line := trimLine(lines[i].text)
		if line == "" || startsIndented(line) {
			continue
		}
		key, value, ok := splitKeyValue(line)
		if ok && key == wanted {
			return i, value, true
		}
	}
	return -1, "", false
}

func replaceLineText(original, replacement string) string {
	switch {
	case strings.HasSuffix(original, "\r\n"):
		return replacement + "\r\n"
	case strings.HasSuffix(original, "\n"):
		return replacement + "\n"
	default:
		return replacement
	}
}

func startsIndented(line string) bool {
	return strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t")
}

func appendDedup(dst []string, values ...string) []string {
	seen := make(map[string]struct{}, len(dst))
	for _, value := range dst {
		seen[value] = struct{}{}
	}
	for _, value := range values {
		if value == "" {
			continue
		}
		if _, ok := seen[value]; ok {
			continue
		}
		seen[value] = struct{}{}
		dst = append(dst, value)
	}
	return dst
}

File: src/wiki/frontmatter.rs
use super::FrontMatter;
use crate::text::{fence_start, normalize_preview_line, scan_lines, starts_indented, trim_line};

pub(crate) fn parse_front_matter(content: &str) -> FrontMatter {
    let lines = scan_lines(content);
    if lines.first().map(|l| trim_line(l.0)) != Some("---") {
        return FrontMatter::default();
    }
    let Some(end_idx) = lines
        .iter()
        .enumerate()
        .skip(1)
        .find(|(_, l)| trim_line(l.0) == "---")
        .map(|(i, _)| i)
    else {
        return FrontMatter::default();
    };
    let mut fm = FrontMatter {
        present: true,
        body_offset: lines[end_idx].2,
        ..Default::default()
    };
    let mut i = 1;
    while i < end_idx {
        let line = trim_line(lines[i].0);
        if line.is_empty() || starts_indented(line) {
            i += 1;
            continue;
        }
        if let Some((key, value)) = split_key_value(line) {
            match key {
                "title" => fm.title = trim_yaml_scalar(value).to_string(),
                "aliases" => {
                    let (aliases, next) = parse_aliases(&lines, i, end_idx, value);
                    for alias in aliases {
                        if !alias.is_empty() && !fm.aliases.contains(&alias) {
                            fm.aliases.push(alias);
                        }
                    }
                    i = next;
                }
                _ => {}
            }
        }
        i += 1;
    }
    fm
}

fn parse_aliases(
    lines: &[(&str, usize, usize)],
    start: usize,
    end: usize,
    value: &str,
) -> (Vec<String>, usize) {
    let trimmed = value.trim();
    let mut next = start;
    if trimmed.is_empty() {
        let mut aliases = Vec::new();
        for (i, line) in lines.iter().enumerate().take(end).skip(start + 1) {
            let text = trim_line(line.0);
            if text.is_empty() {
                next = i;
                continue;
            }
            if !starts_indented(text) {
                break;
            }
            let item = text.trim();
            if let Some(rest) = item.strip_prefix("- ") {
                aliases.push(trim_yaml_scalar(rest.trim()).to_string());
            }
            next = i;
        }
        return (aliases, next);
    }
    if trimmed.starts_with('[') && trimmed.ends_with(']') {
        return (parse_inline_list(trimmed), next);
    }
    (vec![trim_yaml_scalar(trimmed).to_string()], next)
}

fn parse_inline_list(value: &str) -> Vec<String> {
    let inner = value[1..value.len() - 1].trim();
    if inner.is_empty() {
        return Vec::new();
    }
    let mut items = Vec::new();
    let mut current = String::new();
    let mut quote = None;
    for ch in inner.chars() {
        if quote.is_none() && (ch == '\'' || ch == '"') {
            quote = Some(ch);
            current.push(ch);
        } else if quote == Some(ch) {
            quote = None;
            current.push(ch);
        } else if quote.is_none() && ch == ',' {
            items.push(trim_yaml_scalar(current.trim()).to_string());
            current.clear();
        } else {
            current.push(ch);
        }
    }
    if !current.is_empty() {
        items.push(trim_yaml_scalar(current.trim()).to_string());
    }
    items
}

pub(crate) fn first_preview_line(content: &str) -> String {
    let fm = parse_front_matter(content);
    let mut in_fence = false;
    let mut fence_marker = '\0';
    let mut fence_width = 0usize;
    for (line, _, _) in scan_lines(&content[fm.body_offset..]) {
        let trimmed = trim_line(line).trim();
        if let Some((marker, width)) = fence_start(trimmed) {
            if in_fence && marker == fence_marker && width >= fence_width {
                in_fence = false;
            } else if !in_fence {
                in_fence = true;
                fence_marker = marker;
                fence_width = width;
            }
            continue;
        }
        if in_fence || trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        return normalize_preview_line(trimmed);
    }
    String::new()
}

pub(crate) fn update_front_matter_title(
    content: &str,
    old_title: &str,
    new_title: &str,
) -> (String, bool) {
    let lines = scan_lines(content);
    if lines.first().map(|l| trim_line(l.0)) != Some("---") {
        return (content.to_string(), false);
    }
    let Some(end_idx) = lines
        .iter()
        .enumerate()
        .skip(1)
        .find(|(_, l)| trim_line(l.0) == "---")
        .map(|(i, _)| i)
    else {
        return (content.to_string(), false);
    };
    let mut out = String::new();
    let mut changed = false;
    for (i, (line, _, _)) in lines.iter().enumerate() {
        if i > 0 && i < end_idx && !starts_indented(trim_line(line)) {
            if let Some((key, value)) = split_key_value(trim_line(line)) {
                if key == "title" && trim_yaml_scalar(value) == old_title {
                    out.push_str(&replace_line_text(
                        line,
                        &format!("title: {}", format_scalar_like(value, new_title)),
                    ));
                    changed = true;
                    continue;
                }
            }
        }
        out.push_str(line);
    }
    if changed {
        (out, true)
    } else {
        (content.to_string(), false)
    }
}

fn split_key_value(line: &str) -> Option<(&str, &str)> {
    let idx = line.find(':')?;
    Some((line[..idx].trim(), line[idx + 1..].trim()))
}

fn trim_yaml_scalar(value: &str) -> &str {
    let value = value.trim();
    if value.len() >= 2
        && ((value.starts_with('"') && value.ends_with('"'))
            || (value.starts_with('\'') && value.ends_with('\'')))
    {
        &value[1..value.len() - 1]
    } else {
        value
    }
}

fn format_scalar_like(original: &str, replacement: &str) -> String {
    let original = original.trim();
    if original.len() >= 2 && original.starts_with('"') && original.ends_with('"') {
        format!("\"{replacement}\"")
    } else if original.len() >= 2 && original.starts_with('\'') && original.ends_with('\'') {
        format!("'{replacement}'")
    } else {
        replacement.to_string()
    }
}

fn replace_line_text(original: &str, replacement: &str) -> String {
    if original.ends_with("\r\n") {
        format!("{replacement}\r\n")
    } else if original.ends_with('\n') {
        format!("{replacement}\n")
    } else {
        replacement.to_string()
    }
}
